src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs

```diff
@@ -565,6 +564,17 @@ internal void Configure()
                 {
                     InitializeConstructorParameters(GetParameterInfoValues(), sourceGenMode: Options.SerializerContext != null);
                 }
+
+                // Check for misuse of required properties
+                // Ideally this check should be done during property.EnsureConfigured but at that point we have no knowledge about parametrized constructors
+                // and we cannot check if property is read-only correctly
+                foreach (KeyValuePair<string, JsonPropertyInfo> jsonPropertyInfoKv in PropertyCache.List)
```

Comment:
I think this check could be moved inside the `InitializePropertyCache` method to avoid iterating over the items yet again.

Reply:
Unfortunately not in current state because we initialize constructor parameter cache after InitializePropertyCache is called which means we can't check for that correctly in here because we don't know which ones are initialized as part parametrized constructor. It would need likely quite a bit of refactoring to move this logic - probably makes more sense to do that refactoring as part of adding parametrized constructors to the public contract.